aten/src/ATen/native/cpu/UnaryOpsKernel.cpp

```diff
@@ -371,6 +371,37 @@ static void polygamma_kernel(TensorIterator& iter, int64_t n) {
   }
 }
 
+static void nan_to_num_kernel(
+    TensorIterator& iter,
+    c10::optional<double> nan,
+    c10::optional<double> pos_inf,
+    c10::optional<double> neg_inf) {
+  if (c10::isIntegralType(iter.dtype(), true)) {
+    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "nan_to_num_cuda", [&]() {
+      cpu_kernel(iter, [=](scalar_t a) -> scalar_t { return a; });
+    });
+  } else {
+    AT_DISPATCH_FLOATING_TYPES_AND(kHalf, iter.dtype(), "nan_to_num", [&]() {
+      scalar_t nan_replacement = static_cast<scalar_t>(nan.value_or(0.));
+      scalar_t pos_inf_replacement = pos_inf.has_value()
+          ? static_cast<scalar_t>(pos_inf.value())
+          : std::numeric_limits<scalar_t>::max();
+      scalar_t neg_inf_replacement = neg_inf.has_value()
+          ? static_cast<scalar_t>(neg_inf.value())
+          : std::numeric_limits<scalar_t>::lowest();
+
+      cpu_kernel(iter, [=](scalar_t a) -> scalar_t {
+        return at::_isfinite(a)
```

Comment:
This check makes conceptual sense, although I wonder if isfinite is performing all the below checks, anyway, so gating those checks on it is redundant. Maybe the compiler is clever enough to store the result, though.

Reply:
Yeah. I was wondering the same. But this way seemed much easier to grasp.

Reply:
You are indeed correct that this wasn't the best way to go about it (as per the benchmark). Have updated the kernel implementation now. Thanks!